source/common/quic/envoy_quic_server_session.cc

```diff
@@ -82,6 +82,9 @@ void EnvoyQuicServerSession::setUpRequestDecoder(EnvoyQuicServerStream& stream)
 void EnvoyQuicServerSession::OnConnectionClosed(const quic::QuicConnectionCloseFrame& frame,
                                                 quic::ConnectionCloseSource source) {
   quic::QuicServerSessionBase::OnConnectionClosed(frame, source);
+  if (quic_stats_.has_value()) {
```

Comment:
Under what circumstances would quic_stats_ not be set?

We *do* have other examples where there are optional blocks of stats, but those are runtime-feature enabled for extra stats detail.

I think in your usage model this will always be set except for tests. Can we have more consistent behavior where we always set the stats-ref? In other words, make QuicStats be a  required ctor arg for EnvoyQuicServerSession?

Reply:
Just like other Optref stats(https://github.com/envoyproxy/envoy/blob/main/source/common/quic/quic_filter_manager_connection_impl.h#L167), this field is not initialized when the object is created. It will be initialized in QuicHttpServerConnectionImpl constructor (https://github.com/envoyproxy/envoy/blob/main/source/common/quic/codec_impl.cc#L29).
From the current flow, QuicStats can't be in the constructor and be initialized at the very beginning.

There have been assertions that those other OptRef stats must be initialized before creating a new stream https://github.com/envoyproxy/envoy/blob/26a60c2c71f8af4cbaa95f59706817174c24e735/source/common/quic/envoy_quic_server_session.cc#L45 and invoke Envoy bug if the field is not initialized.

I intended to do the same in EnvoyQuicServerSession::OnConnectionClosed(). But unlike the assertions in CreateIncomingStream() where the session is guaranteed to be active and valid, OnConnectionClosed() could be called before the HCM setup.

For example, we could receive a malformed client hello on the wire, which would create a EnvoyQuicServerSession and immediately close it because of protocol violation.

Reply:
Just like other Optref [stats](https://github.com/envoyproxy/envoy/blob/main/source/common/quic/quic_filter_manager_connection_impl.h#L167), this field is not initialized when the object is created. It will be initialized in QuicHttpServerConnectionImpl [constructor](https://github.com/envoyproxy/envoy/blob/main/source/common/quic/codec_impl.cc#L29).
From the current flow, QuicStats can't be in the constructor and be initialized at the very beginning.

There have been assertions that those other OptRef stats must be initialized before creating a new [stream](https://github.com/envoyproxy/envoy/blob/26a60c2c71f8af4cbaa95f59706817174c24e735/source/common/quic/envoy_quic_server_session.cc#L45) and invoke Envoy bug if the field is not initialized.

I intended to do the same in EnvoyQuicServerSession::OnConnectionClosed(). But unlike the assertions in CreateIncomingStream() where the session is guaranteed to be initialized by HCM, OnConnectionClosed() could be called before the HCM setup.

One example is that in EnvoyQuicDispatcherTest.CloseConnectionDuringFilterInstallation, the connection is closed during processing CHLO. 
I'm not sure if that is a valid case in production. But this does mean that if we let the HCM own quic_stats_, we will miss out logging stats when the connection is closed before HCM filter is initialized. Tagging @alyssawilk as Alyssa knows more on this QUIC integration.

Reply:
There's a difference. The other OptRef stats are optional due to config. These are not. I'm trying to suss out why you need a required object to be declared as Optional, as I suspect it's an indication something else is not quite right (see my other comment).

I see that EnvoyQuicConnection is instantiated by EnvoyQuicDispatcher. I'm inferring there's one of those per thread, so it's not a good place for QuicStats to be owned, though it's fine for it to have a reference.

I also see that it is created in `EnvoyQuicDIspatcher::CreateQuicSession`. EnvoyQuicDispatcher has member variables so I think we can supply QuicStats as a ctor arg. Let's go up-stack again.

That gets created from an ActiveQuicListener, which gets created from ActiveQuicListenerFactory, and that's created by Server::ListenerImpl::buildUdpListenerFactory. Cool.

ListenerImpl is created from ListenerManagerFactory. I think ListenerManagerFactory can own the QuicImpl, and pass it down through the ctor/chain so it's available where EnvoyQuicConnection is instantiated.

I think this is a better place because I want there to be exactly one of them in a Server, and it should be available unconditionally whenever a EnvoyQuicConnection is instantiated.

The reason is that there can be many different listener scopes, for different ports that Envoy is listening on. But there should be only one QuicStats object (actually a better name for that structure would be QuicStatNames), since it doesn't actually hold any counters. It just holds their lazily symbolized names.

Maybe what I'm saying is that HCM is not quite right because there's one of them for each listener? Not sure. In any case the order-of-construction seems wrong.

The reason the plumbing needs to be a little more complex than what you have here is to get the ownership/sharing model right.